File: kotlin/src/main/kotlin/foundation/zoobc/zbc/Encryption.kt
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
package foundation.zoobc.zbc

import java.math.BigInteger
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * Sealed transaction messages (spec/signing.md section 8): "ZBE1" || libsodium sealed box to the recipient's
 * Ed25519 key converted to X25519. X25519 (RFC 7748) on BigInteger, HSalsa20, XSalsa20 and Poly1305 written
 * here: the JDK has X25519 but neither Salsa20 nor Poly1305, and one self-contained rendering of the spec is
 * easier to check than a mix.
 */
object Encryption {
    /** The 4-byte marker in front of a sealed message field. */
    @JvmField val SEALED_MAGIC: ByteArray = "ZBE1".toByteArray()
    /** How much longer a sealed field is than its plaintext: marker 4 + ephemeral key 32 + tag 16. */
    const val SEALED_OVERHEAD = 52

    private val P = BigInteger.TWO.pow(255).subtract(BigInteger.valueOf(19))
    private val A24 = BigInteger.valueOf(121665)
    private val MASK255 = BigInteger.TWO.pow(255).subtract(BigInteger.ONE)

    private fun fromLe(b: ByteArray): BigInteger = BigInteger(1, b.reversedArray())
    private fun toLe(v: BigInteger, n: Int): ByteArray {
        val be = v.toByteArray(); val out = ByteArray(n)
        var i = 0; var j = be.size - 1
        while (i < n && j >= 0) { out[i++] = be[j--] }
        return out
    }
    private fun clamp(k: ByteArray): ByteArray = k.copyOf().also { it[0] = (it[0].toInt() and 248).toByte(); it[31] = ((it[31].toInt() and 127) or 64).toByte() }

    /** X25519(scalar, u): 32 bytes. The scalar is clamped as the RFC prescribes. */
    @JvmStatic
    fun x25519(scalar: ByteArray, u: ByteArray): ByteArray {
        require(scalar.size == 32 && u.size == 32) { "X25519 takes 32-byte inputs" }
        val s = fromLe(clamp(scalar)); val x1 = fromLe(u).and(MASK255)
        var x2 = BigInteger.ONE; var z2 = BigInteger.ZERO; var x3 = x1; var z3 = BigInteger.ONE; var swap = 0
        for (t in 254 downTo 0) {
            val kt = if (s.testBit(t)) 1 else 0
            swap = swap xor kt
            if (swap == 1) { val tx = x2; x2 = x3; x3 = tx; val tz = z2; z2 = z3; z3 = tz }
            swap = kt
            val a = x2.add(z2).mod(P); val aa = a.multiply(a).mod(P); val b = x2.subtract(z2).mod(P); val bb = b.multiply(b).mod(P); val e = aa.subtract(bb).mod(P)
            val c = x3.add(z3).mod(P); val d = x3.subtract(z3).mod(P); val da = d.multiply(a).mod(P); val cb = c.multiply(b).mod(P)
            x3 = da.add(cb).mod(P).let { it.multiply(it) }.mod(P); z3 = x1.multiply(da.subtract(cb).mod(P).let { it.multiply(it) }).mod(P)
            x2 = aa.multiply(bb).mod(P); z2 = e.multiply(aa.add(A24.multiply(e)).mod(P)).mod(P)
        }
        if (swap == 1) { val tx = x2; x2 = x3; x3 = tx; val tz = z2; z2 = z3; z3 = tz }
        return toLe(x2.multiply(z2.modInverse(P)).mod(P), 32)
    }

    /** The X25519 public key of a scalar: X25519(scalar, 9). */
    @JvmStatic fun x25519Base(scalar: ByteArray): ByteArray = x25519(scalar, ByteArray(32).also { it[0] = 9 })

    /** Ed25519 public key -> X25519 public key: u = (1 + y) / (1 - y) mod p. */
    @JvmStatic
    fun ed25519PublicKeyToX25519(pk: ByteArray): ByteArray {
        require(pk.size == 32) { "public key must be 32 bytes" }
        val y = fromLe(pk).and(MASK255)
        val den = BigInteger.ONE.subtract(y).mod(P)
        require(den.signum() != 0) { "public key has no X25519 form" }
        return toLe(BigInteger.ONE.add(y).multiply(den.modInverse(P)).mod(P), 32)
    }

    /** Ed25519 seed -> X25519 secret key: the clamped first half of SHA-512(seed). */
    @JvmStatic fun ed25519SeedToX25519(seed: ByteArray): ByteArray = clamp(MessageDigest.getInstance("SHA-512").digest(seed).copyOf(32))

    private val SIGMA = intArrayOf(0x61707865, 0x3320646e, 0x79622d32, 0x6b206574)   // "expand 32-byte k"
    private val QUARTERS = arrayOf(intArrayOf(0, 4, 8, 12), intArrayOf(5, 9, 13, 1), intArrayOf(10, 14, 2, 6), intArrayOf(15, 3, 7, 11),
        intArrayOf(0, 1, 2, 3), intArrayOf(5, 6, 7, 4), intArrayOf(10, 11, 8, 9), intArrayOf(15, 12, 13, 14))

    private fun le32(b: ByteArray, off: Int): Int = (b[off].toInt() and 0xff) or ((b[off + 1].toInt() and 0xff) shl 8) or ((b[off + 2].toInt() and 0xff) shl 16) or ((b[off + 3].toInt() and 0xff) shl 24)
    private fun put32(out: ByteArray, off: Int, v: Int) { out[off] = v.toByte(); out[off + 1] = (v ushr 8).toByte(); out[off + 2] = (v ushr 16).toByte(); out[off + 3] = (v ushr 24).toByte() }
    private fun salsaRounds(x: IntArray) {
        repeat(10) {
            for (q in QUARTERS) {
                val (a, b, c, d) = q
                x[b] = x[b] xor Integer.rotateLeft(x[a] + x[d], 7); x[c] = x[c] xor Integer.rotateLeft(x[b] + x[a], 9)
                x[d] = x[d] xor Integer.rotateLeft(x[c] + x[b], 13); x[a] = x[a] xor Integer.rotateLeft(x[d] + x[c], 18)
            }
        }
    }
    private fun salsaState(key: ByteArray): IntArray {
        val x = IntArray(16)
        x[0] = SIGMA[0]; x[5] = SIGMA[1]; x[10] = SIGMA[2]; x[15] = SIGMA[3]
        for (i in 0 until 4) { x[1 + i] = le32(key, 4 * i); x[11 + i] = le32(key, 16 + 4 * i) }
        return x
    }

    /** HSalsa20(key 32, input 16) -> 32 bytes. */
    @JvmStatic
    fun hsalsa20(key: ByteArray, input: ByteArray): ByteArray {
        val x = salsaState(key)
        for (i in 0 until 4) x[6 + i] = le32(input, 4 * i)
        salsaRounds(x)
        val out = ByteArray(32)
        intArrayOf(0, 5, 10, 15, 6, 7, 8, 9).forEachIndexed { i, w -> put32(out, 4 * i, x[w]) }
        return out
    }

    private fun salsa20Block(key: ByteArray, nonce8: ByteArray, counter: Long): ByteArray {
        val x0 = salsaState(key)
        x0[6] = le32(nonce8, 0); x0[7] = le32(nonce8, 4); x0[8] = counter.toInt(); x0[9] = (counter ushr 32).toInt()
        val x = x0.copyOf(); salsaRounds(x)
        val out = ByteArray(64)
        for (i in 0 until 16) put32(out, 4 * i, x[i] + x0[i])
        return out
    }

    /** The XSalsa20 keystream of a 32-byte key and a 24-byte nonce. */
    @JvmStatic
    fun xsalsa20Stream(key: ByteArray, nonce24: ByteArray, length: Int): ByteArray {
        val sub = hsalsa20(key, nonce24.copyOfRange(0, 16)); val nonce8 = nonce24.copyOfRange(16, 24)
        val out = ByteArray(length); var off = 0; var c = 0L
        while (off < length) { val blk = salsa20Block(sub, nonce8, c++); val n = minOf(64, length - off); System.arraycopy(blk, 0, out, off, n); off += n }
        return out
    }

    /** Poly1305 one-time authenticator (RFC 8439 section 2.5). */
    @JvmStatic
    fun poly1305(key32: ByteArray, msg: ByteArray): ByteArray {
        val r = fromLe(key32.copyOfRange(0, 16)).and(BigInteger("0ffffffc0ffffffc0ffffffc0fffffff", 16)); val s = fromLe(key32.copyOfRange(16, 32))
        val p = BigInteger.TWO.pow(130).subtract(BigInteger.valueOf(5))
        var acc = BigInteger.ZERO
        var i = 0
        while (i < msg.size) {
            val n = minOf(16, msg.size - i)
            acc = acc.add(fromLe(msg.copyOfRange(i, i + n)).setBit(8 * n)).multiply(r).mod(p)
            i += n
        }
        return toLe(acc.add(s).and(BigInteger.TWO.pow(128).subtract(BigInteger.ONE)), 16)
    }

    /** crypto_secretbox_easy: tag (16) || ciphertext. */
    @JvmStatic
    fun secretbox(key: ByteArray, nonce: ByteArray, plaintext: ByteArray): ByteArray {
        val stream = xsalsa20Stream(key, nonce, 32 + plaintext.size)
        val c = ByteArray(plaintext.size) { (plaintext[it].toInt() xor stream[32 + it].toInt()).toByte() }
        return poly1305(stream.copyOfRange(0, 32), c) + c
    }

    /** crypto_secretbox_open_easy: the plaintext, or null when the tag does not verify. */
    @JvmStatic
    fun secretboxOpen(key: ByteArray, nonce: ByteArray, boxed: ByteArray): ByteArray? {
        if (boxed.size < 16) return null
        val c = boxed.copyOfRange(16, boxed.size); val stream = xsalsa20Stream(key, nonce, 32 + c.size)
        if (!MessageDigest.isEqual(poly1305(stream.copyOfRange(0, 32), c), boxed.copyOfRange(0, 16))) return null
        return ByteArray(c.size) { (c[it].toInt() xor stream[32 + it].toInt()).toByte() }
    }

    private fun boxKey(sk: ByteArray, pk: ByteArray): ByteArray = hsalsa20(x25519(sk, pk), ByteArray(16))
    private fun sealedNonce(ephemeralPk: ByteArray, recipientPkX: ByteArray): ByteArray = Blake2b.digest(ephemeralPk + recipientPkX, 24)

    /** True when the field starts with the marker (it may still fail to open). */
    @JvmStatic fun isSealed(field: ByteArray): Boolean = field.size >= 4 && field.copyOfRange(0, 4).contentEquals(SEALED_MAGIC)

    /** Seal `plaintext` to the recipient's 32-byte Ed25519 public key. The ephemeral secret key is random unless given (tests). */
    @JvmStatic
    @JvmOverloads
    fun seal(plaintext: ByteArray, recipientPublicKey: ByteArray, ephemeralSecretKey: ByteArray? = null): ByteArray {
        val rpk = ed25519PublicKeyToX25519(recipientPublicKey)
        val esk = ephemeralSecretKey ?: ByteArray(32).also { SecureRandom().nextBytes(it) }
        require(esk.size == 32) { "ephemeral secret key must be 32 bytes" }
        val epk = x25519Base(esk)
        return SEALED_MAGIC + epk + secretbox(boxKey(esk, rpk), sealedNonce(epk, rpk), plaintext)
    }

    /** Open a sealed field with the recipient's 32-byte Ed25519 seed: the plaintext, or null when it is not a sealed message or the key does not open it. */
    @JvmStatic
    fun openSealed(field: ByteArray, recipientSeed: ByteArray): ByteArray? {
        if (!isSealed(field) || field.size < SEALED_OVERHEAD || recipientSeed.size != 32) return null
        val sk = ed25519SeedToX25519(recipientSeed); val pk = x25519Base(sk); val epk = field.copyOfRange(4, 36)
        return secretboxOpen(boxKey(sk, epk), sealedNonce(epk, pk), field.copyOfRange(36, field.size))
    }
}

File: swift/Sources/ZBC/Encryption.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci

import Crypto
import Foundation

/// Sealed transaction messages (spec/signing.md section 8): "ZBE1" || libsodium sealed box to the recipient's
/// Ed25519 key converted to X25519. X25519 and SHA-512 come from CryptoKit / swift-crypto; the key conversion
/// uses the package's own field arithmetic; HSalsa20, XSalsa20 and Poly1305 are written here (CryptoKit has none).
public enum Encryption {
    /// The 4-byte marker in front of a sealed message field.
    public static let sealedMagic: [UInt8] = Array("ZBE1".utf8)
    /// How much longer a sealed field is than its plaintext: marker 4 + ephemeral key 32 + tag 16.
    public static let sealedOverhead = 52

    /// Ed25519 public key -> X25519 public key of the same point: u = (1 + y) / (1 - y) mod p.
    public static func ed25519PublicKeyToX25519(_ pk: [UInt8]) -> [UInt8]? {
        guard pk.count == 32 else { return nil }
        var yb = pk
        yb[31] &= 0x7f
        let y = Ed25519.reduceP(Ed25519.Big(le: yb))
        let one = Ed25519.Big(1)
        let den = Ed25519.fsub(one, y)
        if den.isZero { return nil }
        let u = Ed25519.fmul(Ed25519.fadd(one, y), Ed25519.finv(den))
        return u.leBytes(32)
    }

    /// Ed25519 seed -> X25519 secret key: the clamped first half of SHA-512(seed).
    public static func ed25519SeedToX25519(_ seed: [UInt8]) -> [UInt8] {
        var sk = Array(Array(SHA512.hash(data: seed)).prefix(32))
        sk[0] &= 248
        sk[31] &= 127
        sk[31] |= 64
        return sk
    }

    /// The X25519 public key of a scalar: X25519(scalar, 9).
    public static func x25519Base(_ scalar: [UInt8]) -> [UInt8]? {
        guard let k = try? Curve25519.KeyAgreement.PrivateKey(rawRepresentation: scalar) else { return nil }
        return Array(k.publicKey.rawRepresentation)
    }

    static func x25519(_ scalar: [UInt8], _ u: [UInt8]) -> [UInt8]? {
        guard let k = try? Curve25519.KeyAgreement.PrivateKey(rawRepresentation: scalar),
              let p = try? Curve25519.KeyAgreement.PublicKey(rawRepresentation: u),
              let shared = try? k.sharedSecretFromKeyAgreement(with: p) else { return nil }
        return shared.withUnsafeBytes { Array($0) }
    }

    private static let sigma: [UInt32] = [0x61707865, 0x3320646e, 0x79622d32, 0x6b206574]   // "expand 32-byte k"
    private static let quarters: [[Int]] = [[0, 4, 8, 12], [5, 9, 13, 1], [10, 14, 2, 6], [15, 3, 7, 11], [0, 1, 2, 3], [5, 6, 7, 4], [10, 11, 8, 9], [15, 12, 13, 14]]

    private static func le32(_ b: [UInt8], _ off: Int) -> UInt32 {
        return UInt32(b[off]) | (UInt32(b[off + 1]) << 8) | (UInt32(b[off + 2]) << 16) | (UInt32(b[off + 3]) << 24)
    }
    private static func put32(_ out: inout [UInt8], _ off: Int, _ v: UInt32) {
        out[off] = UInt8(v & 0xff); out[off + 1] = UInt8((v >> 8) & 0xff); out[off + 2] = UInt8((v >> 16) & 0xff); out[off + 3] = UInt8((v >> 24) & 0xff)
    }
    @inline(__always) private static func rotl(_ v: UInt32, _ c: UInt32) -> UInt32 { (v << c) | (v >> (32 - c)) }
    private static func salsaRounds(_ x: inout [UInt32]) {
        for _ in 0..<10 {
            for q in quarters {
                let a = q[0], b = q[1], c = q[2], d = q[3]
                x[b] ^= rotl(x[a] &+ x[d], 7)
                x[c] ^= rotl(x[b] &+ x[a], 9)
                x[d] ^= rotl(x[c] &+ x[b], 13)
                x[a] ^= rotl(x[d] &+ x[c], 18)
            }
        }
    }
    private static func salsaState(_ key: [UInt8]) -> [UInt32] {
        var x = [UInt32](repeating: 0, count: 16)
        x[0] = sigma[0]; x[5] = sigma[1]; x[10] = sigma[2]; x[15] = sigma[3]
        for i in 0..<4 { x[1 + i] = le32(key, 4 * i); x[11 + i] = le32(key, 16 + 4 * i) }
        return x
    }

    /// HSalsa20(key 32, input 16) -> 32 bytes.
    public static func hsalsa20(_ key: [UInt8], _ input: [UInt8]) -> [UInt8] {
        var x = salsaState(key)
        for i in 0..<4 { x[6 + i] = le32(input, 4 * i) }
        salsaRounds(&x)
        var out = [UInt8](repeating: 0, count: 32)
        let words = [0, 5, 10, 15, 6, 7, 8, 9]
        for i in 0..<8 { put32(&out, 4 * i, x[words[i]]) }
        return out
    }

    private static func salsa20Block(_ key: [UInt8], _ nonce8: [UInt8], _ counter: UInt64) -> [UInt8] {
        var x0 = salsaState(key)
        x0[6] = le32(nonce8, 0); x0[7] = le32(nonce8, 4)
        x0[8] = UInt32(truncatingIfNeeded: counter); x0[9] = UInt32(truncatingIfNeeded: counter >> 32)
        var x = x0
        salsaRounds(&x)
        var out = [UInt8](repeating: 0, count: 64)
        for i in 0..<16 { put32(&out, 4 * i, x[i] &+ x0[i]) }
        return out
    }

    /// The XSalsa20 keystream of a 32-byte key and a 24-byte nonce.
    public static func xsalsa20Stream(_ key: [UInt8], _ nonce24: [UInt8], _ length: Int) -> [UInt8] {
        let sub = hsalsa20(key, Array(nonce24[0..<16]))
        let nonce8 = Array(nonce24[16..<24])
        var out: [UInt8] = []
        out.reserveCapacity(length + 64)
        var c: UInt64 = 0
        while out.count < length {
            out.append(contentsOf: salsa20Block(sub, nonce8, c))
            c += 1
        }
        return Array(out.prefix(length))
    }

    /// Poly1305 one-time authenticator (RFC 8439 section 2.5).
    public static func poly1305(_ key32: [UInt8], _ msg: [UInt8]) -> [UInt8] {
        var rb = Array(key32[0..<16])
        rb[3] &= 15; rb[7] &= 15; rb[11] &= 15; rb[15] &= 15; rb[4] &= 252; rb[8] &= 252; rb[12] &= 252
        let r = Ed25519.Big(le: rb)
        let s = Ed25519.Big(le: Array(key32[16..<32]))
        let p = Ed25519.Big(1).shiftedLeft(130) - Ed25519.Big(5)
        var acc = Ed25519.Big([])
        var i = 0
        while i < msg.count {
            let n = min(16, msg.count - i)
            var block = Array(msg[i..<(i + n)])
            block.append(1)
            acc = ((acc + Ed25519.Big(le: block)) * r).mod(p)
            i += n
        }
        return (acc + s).leBytes(16)
    }

    /// crypto_secretbox_easy: tag (16) || ciphertext.
    public static func secretbox(_ key: [UInt8], _ nonce: [UInt8], _ plaintext: [UInt8]) -> [UInt8] {
        let stream = xsalsa20Stream(key, nonce, 32 + plaintext.count)
        var c = [UInt8](repeating: 0, count: plaintext.count)
        for i in 0..<plaintext.count { c[i] = plaintext[i] ^ stream[32 + i] }
        return poly1305(Array(stream[0..<32]), c) + c
    }

    /// crypto_secretbox_open_easy: the plaintext, or nil when the tag does not verify.
    public static func secretboxOpen(_ key: [UInt8], _ nonce: [UInt8], _ boxed: [UInt8]) -> [UInt8]? {
        guard boxed.count >= 16 else { return nil }
        let c = Array(boxed[16...])
        let stream = xsalsa20Stream(key, nonce, 32 + c.count)
        let tag = poly1305(Array(stream[0..<32]), c)
        var diff: UInt8 = 0
        for i in 0..<16 { diff |= tag[i] ^ boxed[i] }
        if diff != 0 { return nil }
        var out = [UInt8](repeating: 0, count: c.count)
        for i in 0..<c.count { out[i] = c[i] ^ stream[32 + i] }
        return out
    }

    private static func boxKey(_ sk: [UInt8], _ pk: [UInt8]) -> [UInt8]? {
        guard let shared = x25519(sk, pk) else { return nil }
        return hsalsa20(shared, [UInt8](repeating: 0, count: 16))
    }
    private static func sealedNonce(_ ephemeralPk: [UInt8], _ recipientPkX: [UInt8]) -> [UInt8] {
        return Blake2b.hash(ephemeralPk + recipientPkX, outLen: 24)
    }

    /// True when the field starts with the marker (it may still fail to open).
    public static func isSealed(_ field: [UInt8]) -> Bool { field.count >= 4 && Array(field[0..<4]) == sealedMagic }

    /// Seal `plaintext` to the recipient's 32-byte Ed25519 public key. The ephemeral secret key is random unless given (tests).
    public static func seal(_ plaintext: [UInt8], recipientPublicKey: [UInt8], ephemeralSecretKey: [UInt8]? = nil) -> [UInt8]? {
        guard let rpk = ed25519PublicKeyToX25519(recipientPublicKey) else { return nil }
        let esk = ephemeralSecretKey ?? (0..<32).map { _ in UInt8.random(in: 0...255) }
        guard esk.count == 32, let epk = x25519Base(esk), let key = boxKey(esk, rpk) else { return nil }
        let boxed = secretbox(key, sealedNonce(epk, rpk), plaintext)
        return sealedMagic + epk + boxed
    }

    /// Open a sealed field with the recipient's 32-byte Ed25519 seed: the plaintext, or nil when it is not a sealed message or the key does not open it.
    public static func openSealed(_ field: [UInt8], recipientSeed: [UInt8]) -> [UInt8]? {
        guard isSealed(field), field.count >= sealedOverhead, recipientSeed.count == 32 else { return nil }
        let sk = ed25519SeedToX25519(recipientSeed)
        guard let pk = x25519Base(sk) else { return nil }
        let epk = Array(field[4..<36])
        guard let key = boxKey(sk, epk) else { return nil }
        return secretboxOpen(key, sealedNonce(epk, pk), Array(field[36...]))
    }
}
